### PR TITLE
Cleaner export and asset store classes

### PR DESCRIPTION
**Pull request by rowanseymour**
* Adds common base class for all exports removing a lot of duplicated code
* Ensure cell values are converted consistently, e.g. dates as org timezone
* Fixes issue where text values beginning with = cause Excel to think a file is corrupt
* Reworks asset stores so they aren't hardcoded in the assets app. Got rid of `AssetType` so now there's just dynamic asset stores.

### CODE REVIEW

File: temba/contacts/migrations/0051_auto_20170214_1009.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+    dependencies = [
+        ('contacts', '0050_contactgroupcount_is_squashed'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='exportcontactstask',
+            name='org',
+            field=models.ForeignKey(help_text='The organization of the user.', on_delete=django.db.models.deletion.CASCADE, related_name='exportcontactstasks', to='orgs.Org'),
```
Comment (rowanseymour) [inline temba/contacts/migrations/0051_auto_20170214_1009.py]: Migration is just because the related name has changed

File: temba/utils/export.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        """
+        try:
+            start = time.time()
+            self.do_export()
+        finally:
+            elapsed = time.time() - start
+            analytics.track(self.created_by.username, 'temba.%s_latency' % self.EXPORT_NAME, properties=dict(value=elapsed))
+
+            self.is_finished = True
+            self.save(update_fields=('is_finished',))
```
Comment (nicpottier) [inline temba/utils/export.py]: Should we be tracking status as well maybe? Not sure how we are tracking success vs failure right now.

File: temba/utils/export.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    class Meta:
+        abstract = True
+
+
+class TableExporter(object):
+    """
+    Class that abstracts out writing a table of data to a CSV or Excel file. This only works for exports that
+    have a single sheet (as CSV's don't have sheets) but takes care of writing to a CSV in the case
+    where there are more than 256 columns, which Excel doesn't support.
+
```
Comment (ericnewcomer) [inline temba/utils/export.py]: Didn't we move to xlsx writers that don't have these limitations?

File: temba/utils/export.py (hunk 2 above)
Comment (rowanseymour) [inline temba/utils/export.py]: Would be pretty straightforward to change this to `status = (PENDING | PROCESSING | FAILED | FINISHED)`. That sound good?

File: temba/utils/export.py (hunk 3 above)
Comment (rowanseymour) [inline temba/utils/export.py]: Oh I think this number is now 2^14..  @norkans7 ?

File: temba/utils/export.py (hunk 2 above)
Comment (nicpottier) [inline temba/utils/export.py]: Ya, let's do that!

File: temba/flows/tests.py [hunk 4]
```diff
@@ -687,7 +687,7 @@ def test_export_results(self):
                                             c1_run1_last, "Orange", "orange", "orange"], tz)
 
         self.assertExcelRow(sheet_runs, 2, [contact1_run2.contact.uuid, "+250788382382", "Eric", "Devs", c1_run2_first,
-                                            c1_run2_last, "Blue", "blue", " blue "], tz)
```
Comment (norkans7) [inline temba/flows/tests.py]: I think these spaces are testing trailing spaces are removed

File: temba/utils/export.py (hunk 3 above)
Comment (norkans7) [inline temba/utils/export.py]: we export xlsx now MAX_XLS_COLS = 16384

Comment (norkans7) [inline temba/utils/export.py]: more than that we switch to CSV

Comment (rowanseymour) [inline temba/utils/export.py]: Not sure we even want to be exporting something with 16384+ columns. Do we even need to keep the CSV option then?

File: temba/flows/tests.py (hunk 4 above)
Comment (rowanseymour) [inline temba/flows/tests.py]: Looks like currently implementation doesn't remove any spaces, no? I made the new stuff trim whitespace

File: temba/utils/export.py (hunk 3 above)
Comment (norkans7) [inline temba/utils/export.py]: I think we allow exporting results of many flows at once so those can have too many columns

File: temba/flows/tests.py (hunk 4 above)
Comment (norkans7) [inline temba/flows/tests.py]: Ah right, I guess we should not trim that one since it is the text from the user who replied

Comment (rowanseymour) [inline temba/flows/tests.py]: I think you've got it mixed up no? This is now testing that we do trim.

Comment (norkans7) [inline temba/flows/tests.py]: I was confused initially, I guess we need to keep the text from users as it was sent to us, with  whitespace

File: temba/assets/models.py [hunk 5]
```diff
@@ -58,7 +67,7 @@ def resolve(self, user, pk):
         if not user.has_org_perm(asset.org, self.permission):  # pragma: needs cover
             raise AssetAccessDenied()
 
-        if not asset.uuid:
+        if not self.is_asset_ready(asset):
```
Comment (rowanseymour) [inline temba/assets/models.py]: Previously exports were considered ready if they have a UUID. Have made this more explicit and now UUID is always set like other models

File: temba/assets/models.py [hunk 6]
```diff
@@ -129,32 +138,9 @@ def derive_path(self, org, uuid, extension=None):
 
         raise AssetFileNotFound()  # pragma: needs cover
 
+    def is_asset_ready(self, asset):
+        return True
 
-class ContactExportAssetStore(BaseAssetStore):
```
Comment (rowanseymour) [inline temba/assets/models.py]: These move into their respective apps to avoid having dependencies going both ways

File: temba/flows/tests.py [hunk 7]
```diff
@@ -656,10 +656,9 @@ def test_export_results(self):
         self.assertContains(response, "already an export in progress")
 
         # ok, mark that one as finished and try again
-        blocking_export.is_finished = True
-        blocking_export.save()
+        blocking_export.update_status(ExportFlowResultsTask.STATUS_COMPLETE)
 
-        with self.assertNumQueries(49):
+        with self.assertNumQueries(50):
```
Comment (rowanseymour) [inline temba/flows/tests.py]: Extra db hit here is just updating the status to PROCESSING once we start processing the export

File: temba/orgs/views.py [hunk 8]
```diff
@@ -2378,28 +2377,6 @@ def pre_process(self, request, *args, **kwargs):
             num_deleted = self.get_object().clear_caches([cache])
             self.success_message = _("Cleared %s cache for this organization (%d keys)") % (cache.name, num_deleted)
 
-    class Download(SmartTemplateView):
```
Comment (rowanseymour) [inline temba/orgs/views.py]: We left this in for old links to exports... which are now several years old so I think we can ditch this view

File: temba/flows/tests.py (hunk 4 above)
Comment (rowanseymour) [inline temba/flows/tests.py]: Have reverted this so whitespace is kept

File: temba/contacts/models.py [hunk 9]
```diff
@@ -2291,30 +2289,12 @@ def __str__(self):  # pragma: needs cover
...
 
-class ExportContactsTask(SmartModel):
+class ExportContactsTask(BaseExportTask):
+    analytics_key = 'contact_export'
+    email_subject = "Your contacts export is ready"
+    email_template = 'contacts/email/contacts_export_download'
 
-    org = models.ForeignKey(Org, related_name='contacts_exports', help_text=_("The Organization of the user."))
     group = models.ForeignKey(ContactGroup, null=True, related_name='exports', help_text=_("The unique group to export"))
-    task_id = models.CharField(null=True, max_length=64)
```
Comment (rowanseymour) [inline temba/contacts/models.py]: All the export classes have this field but it doesn't appear to have ever been used

**Pull request merged**